File: python/tests/test_eval_routing.py
"""Parity tests for ``eval_exp_tens`` single-multiset centres-path routing through
the shared helper (:func:`mpt._kernel.gaussian_kernel_sum`).

Verifies, on a battery of representative densities and queries, that:

- Default settings (``truncation_sigmas=inf``, ``kernel_precision='double'``)
  produce FP-bit-identical output to a frozen reference implementation
  matching v2.0/v2.1 behaviour.
- ``truncation_sigmas=6`` agrees with the reference to bounded relative
  error (cumulative bound proportional to ``sum(|wJ|) * exp(-18)``).
- ``kernel_precision='single'`` agrees within ~1e-5 relative.
- Per-call kwargs override toolbox defaults.
- Toolbox defaults (set via ``mpt.set_default``) are picked up when
  per-call kwargs are absent.

The reference implementation is a plain broadcast-subtract-exp-sum,
matching what ``_eval_core`` did in v2.0/v2.1 — kept here as a stable
local-frozen check rather than relying on the now-routed production
code.
"""

import math
import zlib

import numpy as np

from mpt._defaults import accuracy_floor_context

# Comparisons against the untruncated reference widen the accuracy floor
# to 1e-300 (as the MATLAB twin test_kernel_truncation.m does), so the
# shipped paths sum essentially exhaustively. Entries below that floor
# are legitimately dropped -- the reference can still emit denormals at
# ~1e-308 -- so atol is tied to the floor being requested.
_FLOOR = 1e-300

# The floor is a per-term cutoff, but a summed value can sit a few
# multiples above it while every term composing it was dropped: at a
# query far from every centre the reference returns ~2e-300 where the
# shipped path returns exactly 0. An atol equal to the floor fails those
# entries by a factor of two or three, so the comparison needs a margin
# above it.
_ATOL = 1e3 * _FLOOR


import pytest

import mpt
from mpt.tensor import build_exp_tens, eval_exp_tens


# ---------------------------------------------------------------------
# Reference (frozen v2.0/v2.1 body)
# ---------------------------------------------------------------------

def _ref_eval(dens, x):
    """Plain broadcast-subtract-exp-sum, no truncation, double precision."""
    # A MaetDensity carries per-attribute vectors (r, sigma, is_rel, ...);
    # this frozen reference is written against the flat single-multiset
    # field names, so take that view rather than indexing the vectors here.
    from mpt._tensor.density import single_multiset_view
    dens = single_multiset_view(dens)
    centres = dens.centres
    w_j = dens.w_j
    sigma = dens.sigma
    r = int(dens.r)
    is_rel = bool(dens.is_rel)
    is_per = bool(dens.is_per)
    period = float(dens.period)

    x = np.asarray(x, dtype=np.float64)
    if x.ndim == 1:
        x = x.reshape(1, -1)

    D = centres[:, :, None] - x[:, None, :]
    if is_rel:
        # A relative density depends on a tuple only through its
        # within-tuple differences, and
        #     sum_i D_i^2 - (sum_i D_i)^2 / r = (1/r) sum_{i<j} (D_i - D_j)^2
        # over the r positions, of which the reduced coordinates carry r - 1
        # with an implicit position 0 at the origin. Periodicity acts on
        # those within-tuple differences, so the wrap belongs inside the
        # pairwise sum -- wrapping each D_i first and then forming the
        # quadratic is a different quantity, and the two part company
        # wherever a within-tuple difference sits beyond half a period.
        D = np.concatenate([np.zeros((1,) + D.shape[1:]), D], axis=0)
        Q = np.zeros(D.shape[1:])
        for i in range(r):
            for j in range(i + 1, r):
                d = D[i] - D[j]
                if is_per:
                    d = np.mod(d + period / 2, period) - period / 2
                Q = Q + d * d
        Q = Q / r
    else:
        if is_per:
            D = np.mod(D + period / 2, period) - period / 2
        Q = np.sum(D * D, axis=0)
    E = np.exp(-Q / (2 * sigma ** 2))
    return w_j @ E


# ---------------------------------------------------------------------
# Fixtures
# ---------------------------------------------------------------------

@pytest.fixture(autouse=True)
def reset_defaults_each_test():
    yield
    mpt.reset_defaults()


# A representative battery: r in {2, 3}, abs and rel, non-periodic
# and periodic, different K, σ.
@pytest.fixture(params=[
    # (description, K, r, is_rel, is_per, period, sigma, n_q)
    ("abs r=2 K=6",        6, 2, False, False, 0.0,    8.0,  30),
    ("rel r=2 K=6",        6, 2, True,  False, 0.0,    8.0,  30),
    ("abs r=3 K=5",        5, 3, False, False, 0.0,   12.0,  20),
    ("rel r=3 K=5",        5, 3, True,  False, 0.0,   12.0,  20),
    ("rel r=3 K=8 σ=30",   8, 3, True,  False, 0.0,   30.0,  50),
    ("rel-per r=3 K=6",    6, 3, True,  True,  1200.0, 25.0, 25),
    ("abs-per r=2 K=6",    6, 2, False, True,  1200.0, 25.0, 25),
])
def density_case(request):
    label, K, r, is_rel, is_per, period, sigma, n_q = request.param
    # Seed from a stable digest of the label, not from hash(): Python
    # salts string hashing per process, so hash(label) gives different
    # data on every run and a failure cannot be reproduced from the
    # reported case name.
    rng = np.random.default_rng(zlib.crc32(label.encode()))
    p = np.sort(rng.uniform(0, 1000, K))
    w = rng.uniform(0.5, 1.5, K)
    dens = build_exp_tens(p, w, sigma, r, is_rel, is_per, period, verbose=False)
    dim = r - 1 if is_rel else r
    if is_per:
        x = rng.uniform(0, period, (dim, n_q))
    else:
        x = rng.uniform(0, 1000, (dim, n_q))
    return label, dens, x


# ---------------------------------------------------------------------
# The shipped paths must match the reference to reduction-order noise
# (~1e-13): the two carry out the same arithmetic in a different order,
# which for periodic configurations differs because the wrap is written
# as D - period * floor(D/period + 0.5) rather than with np.mod.
#
# That holds only because the reference forms the relative quadratic
# from wrapped within-tuple differences, as the shipped paths do. An
# earlier reference wrapped each offset before forming the quadratic,
# which is a different quantity: the two agree near the peak of the
# density and part company in its tail, where a within-tuple difference
# can sit beyond half a period. The disagreement reached 7e-3 relative
# at values ~1e-53, on roughly one draw in three thousand.
# ---------------------------------------------------------------------

def test_default_settings_match_reference(density_case):
    label, dens, x = density_case
    with accuracy_floor_context(1e-300):
        v = eval_exp_tens(dens, x, method='centres', truncation_sigmas=math.inf,
                          verbose=False)
    ref = _ref_eval(dens, x)
    np.testing.assert_allclose(
        v, ref, rtol=1e-12, atol=_ATOL,
        err_msg=f"{label}: divergence from reference exceeds rtol=1e-12"
    )


def test_explicit_inf_matches_reference(density_case):
    label, dens, x = density_case
    with accuracy_floor_context(1e-300):
        v = eval_exp_tens(
            dens, x, method='centres', truncation_sigmas=math.inf,
            kernel_precision='double', verbose=False,
        )
    ref = _ref_eval(dens, x)
    np.testing.assert_allclose(
        v, ref, rtol=1e-12, atol=_ATOL,
        err_msg=f"{label}: divergence from reference exceeds rtol=1e-12"
    )


# ---------------------------------------------------------------------
# Truncation
# ---------------------------------------------------------------------

@pytest.mark.parametrize("k", [4, 5, 6])
def test_truncation_within_bound(density_case, k):
    label, dens, x = density_case
    # Periodic mode falls through to exact: truncation has no effect.
    if dens.is_per:
        pytest.skip("periodic mode currently exact-only")
    v_trunc = eval_exp_tens(
        dens, x, method='centres', truncation_sigmas=k, verbose=False,
    )
    ref = _ref_eval(dens, x)
    # Cumulative bound: sum(|wJ|) * exp(-k^2/2). Add 10x safety.
    weight_mass = float(np.sum(np.abs(dens.w_j)))
    bound = max(1e-12, 10 * weight_mass * math.exp(-k ** 2 / 2))
    # Use absolute error rather than relative — relative blows up when
    # ref values are tiny (queries far from all centres).
    err = np.max(np.abs(v_trunc - ref))
    assert err < bound, \
        f"{label} k={k}: err={err:.3e}, bound={bound:.3e}, " \
        f"weight_mass={weight_mass:.3f}"


def test_truncation_inf_is_exact(density_case):
    """At truncation_sigmas=inf, output is bit-identical to default."""
    label, dens, x = density_case
    v_default = eval_exp_tens(dens, x, method='centres', verbose=False)
    v_inf = eval_exp_tens(
        dens, x, method='centres', truncation_sigmas=math.inf, verbose=False,
    )
    assert np.array_equal(v_default, v_inf), \
        f"{label}: explicit inf vs default not bit-identical"


def test_periodic_truncation_agrees_within_floor(density_case):
    """Periodic mode: culling changes values only below the requested floor.

    Periodic evaluation once fell through to an exact wrapped sum, so
    truncation had no effect and the two calls were bit-identical. Since
    periodic-mode culling shipped, a requested width does cull, so the
    contract is agreement to the amplitude that width admits --
    exp(-k^2/2) at k = 6 -- not bitwise equality.
    """
    label, dens, x = density_case
    if not dens.is_per:
        pytest.skip("only applies to periodic")
    v_default = eval_exp_tens(dens, x, method='centres', verbose=False)
    v_trunc = eval_exp_tens(
        dens, x, method='centres', truncation_sigmas=6, verbose=False,
    )
    np.testing.assert_allclose(
        v_trunc, v_default, rtol=0, atol=10.0 * math.exp(-6.0 ** 2 / 2),
        err_msg=f"{label}: periodic culling exceeds the requested floor",
    )


# ---------------------------------------------------------------------
# Precision
# ---------------------------------------------------------------------

def test_single_precision_within_bound(density_case):
    label, dens, x = density_case
    v_single = eval_exp_tens(
        dens, x, method='centres', kernel_precision='single', verbose=False,
    )
    ref = _ref_eval(dens, x)
    # Single precision: ~1e-7 rel; allow generous margin. Compare on
    # absolute error to avoid blow-up when ref values are tiny.
    weight_mass = float(np.sum(np.abs(dens.w_j)))
    err = np.max(np.abs(v_single - ref))
    # Single-precision exp() carries ~1e-7 rel; absolute error bounded
    # by max kernel value times sum(|wJ|) times 1e-7. Allow 1e-5 of
    # weight mass for safety.
    bound = max(1e-12, 1e-5 * weight_mass)
    assert err < bound, f"{label}: single abs err {err:.3e}, bound {bound:.3e}"


# ---------------------------------------------------------------------
# Global defaults
# ---------------------------------------------------------------------

def test_global_default_picked_up():
    """When per-call kwarg is absent, the helper consults mpt defaults."""
    rng = np.random.default_rng(42)
    p = np.sort(rng.uniform(0, 1000, 6))
    w = rng.uniform(0.5, 1.5, 6)
    sigma = 12.0
    dens = build_exp_tens(p, w, sigma, 3, True, False, 0.0, verbose=False)
    x = rng.uniform(0, 1000, (2, 20))
    ref = _ref_eval(dens, x)

    # With the default width, agreement is to the accuracy floor: the
    # resolved width is k = 7.43, not an untruncated sum.
    with accuracy_floor_context(1e-300):
        v1 = eval_exp_tens(dens, x, method='centres',
                           truncation_sigmas=math.inf, verbose=False)
    np.testing.assert_allclose(v1, ref, rtol=1e-12, atol=_ATOL)

    # Set global default; bit-identical now requires truncation match.
    mpt.set_default(truncation_sigmas=6)
    v2 = eval_exp_tens(dens, x, method='centres', verbose=False)
    weight_mass = float(np.sum(np.abs(dens.w_j)))
    bound = max(1e-12, 10 * weight_mass * math.exp(-18))
    err = np.max(np.abs(v2 - ref))
    assert err < bound


def test_per_call_overrides_global():
    rng = np.random.default_rng(43)
    p = np.sort(rng.uniform(0, 1000, 6))
    w = rng.uniform(0.5, 1.5, 6)
    sigma = 12.0
    dens = build_exp_tens(p, w, sigma, 3, True, False, 0.0, verbose=False)
    x = rng.uniform(0, 1000, (2, 20))
    ref = _ref_eval(dens, x)

    # Set a lax global default.
    mpt.set_default(truncation_sigmas=4)
    # A per-call Inf must still resolve to the accuracy floor, overriding
    # the laxer global width.
    with accuracy_floor_context(1e-300):
        v = eval_exp_tens(
            dens, x, method='centres', truncation_sigmas=math.inf,
            verbose=False,
        )
    np.testing.assert_allclose(v, ref, rtol=1e-12, atol=_ATOL)

File: matlab/tests/test_eval_routing.m
%% test_eval_routing.m — v3 evalExpTens centres-path routing parity
%
%  Mirrors python/tests/test_eval_routing.py. Verifies that the
%  v3 refactor of localEvalSingleMultisetCentres (routing through
%  internal.gaussianKernelSum) is:
%
%    - Within the accuracy-floor truncation bound of the v2.0
%      untruncated reference at default settings (Inf resolves to the
%      ~7.43 sigma floor; kernelPrecision='double').
%    - Within bounded relative error at finite truncationSigmas.
%    - Within ~1e-5 relative at kernelPrecision='single'.
%
%  Standalone-runnable. When invoked from test_mpt.m the existing
%  `results` cell is appended to.

if ~exist('results', 'var')
    results = {};
    standalone_er = true;
    % Defaults isolation when run standalone (when invoked from
    % test_mpt.m the outer wrapper has already isolated defaults).
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults_er
    cleanupDefaults_er = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone_er = false;
end

mptDefaults('reset');
% reset now yields the factory truncationSigmas = 6; pin Inf so the
% "default matches reference" comparisons below exercise the accuracy-
% floor width (Inf resolves to ~7.43 sigma), checked within the
% truncation bound against the untruncated reference.
mptDefaults('truncationSigmas', Inf);

% --- Battery of representative cases ---
% Each row: {label, K, r, isRel, isPer, period, sigma, nQ}
er_cases = { ...
    'abs r=2 K=6',     6, 2, false, false, 0,    8,  30; ...
    'rel r=2 K=6',     6, 2, true,  false, 0,    8,  30; ...
    'abs r=3 K=5',     5, 3, false, false, 0,   12,  20; ...
    'rel r=3 K=5',     5, 3, true,  false, 0,   12,  20; ...
    'rel r=3 K=8 s=30',8, 3, true,  false, 0,   30,  50; ...
    'rel-per r=3 K=6', 6, 3, true,  true,  1200,25,  25; ...
    'abs-per r=2 K=6', 6, 2, false, true,  1200,25,  25; ...
};

for ic = 1:size(er_cases, 1)
    label  = er_cases{ic, 1};
    K      = er_cases{ic, 2};
    r      = er_cases{ic, 3};
    isRel  = er_cases{ic, 4};
    isPer  = er_cases{ic, 5};
    period = er_cases{ic, 6};
    sigma  = er_cases{ic, 7};
    nQ     = er_cases{ic, 8};

    % Deterministic seed per case (label hash)
    rng(sum(double(label)), 'twister');
    p = sort(1000 * rand(K, 1));
    w = 0.5 + rand(K, 1);
    dens = buildExpTens(p, w, sigma, r, isRel, isPer, period, ...
        'verbose', false);
    % Populate heavy fields (Centres, wJ) so the test body can read
    % them. Production code calls internal.ensureExpTensExpensive inside the
    % evalExpTens centres-branch before invoking localEvalSingleMultisetCentres.
    dens = internal.ensureExpTensExpensive(dens);
    if isRel
        dim = r - 1;
    else
        dim = r;
    end
    if isPer
        X = period * rand(dim, nQ);
    else
        X = 1000 * rand(dim, nQ);
    end

    % --- Reference: frozen v2.0 body ---
    ref = local_ref_eval(dens, X);

    % --- Default / explicit Inf: the reference sums untruncated, but Inf
    %     now resolves to the accuracy-floor width (~7.43 sigma), so eval
    %     truncates there. Check against the reference within the same
    %     truncation bound the finite-k cases use below, at the resolved
    %     floor width. Reduction-order differences (e.g. the periodic wrap
    %     retargeted from `mod` to `D - period .* floor(D/period + 0.5)`,
    %     mathematically equivalent but a different FP-op sequence) sit far
    %     below this bound.
    kFloor      = internal.accuracyFloor('resolve', Inf);
    weight_mass = sum(abs(dens.wJ));
    floorBound  = max(1e-12, 10 * weight_mass * exp(-kFloor^2 / 2));

    v_default = evalExpTens(dens, X, 'method', 'centres', 'verbose', false);
    results{end+1, 1} = sprintf('eval_routing: default matches reference (%s)', label); %#ok<*AGROW>
    results{end, 2} = max(abs(v_default(:) - ref(:))) < floorBound;

    % --- Explicit Inf/double: same floor-width truncation bound ---
    v_inf = evalExpTens(dens, X, 'method', 'centres', ...
        'truncationSigmas', Inf, 'kernelPrecision', 'double', ...
        'verbose', false);
    results{end+1, 1} = sprintf('eval_routing: Inf/double matches reference (%s)', label);
    results{end, 2} = max(abs(v_inf(:) - ref(:))) < floorBound;

    % --- truncationSigmas at k=4..6 within cumulative bound ---
    % Non-periodic only: under the truncation contract, periodic mode
    % also truncates (inf resolves to the accuracy-floor width), so the
    % former "periodic ignores truncation" check is obsolete and removed.
    if ~isPer
        for k = [4, 5, 6]
            v_trunc = evalExpTens(dens, X, 'method', 'centres', ...
                'truncationSigmas', k, 'verbose', false);
            weight_mass = sum(abs(dens.wJ));
            bound = max(1e-12, 10 * weight_mass * exp(-k^2 / 2));
            err = max(abs(v_trunc - ref));
            results{end+1, 1} = sprintf(...
                'eval_routing: truncation k=%d within bound (%s)', k, label);
            results{end, 2} = err < bound;
        end
    end

    % --- Single precision within bound ---
    v_single = evalExpTens(dens, X, 'method', 'centres', ...
        'kernelPrecision', 'single', 'verbose', false);
    weight_mass = sum(abs(dens.wJ));
    bound_single = max(1e-12, 1e-5 * weight_mass);
    err_single = max(abs(v_single - ref));
    results{end+1, 1} = sprintf(...
        'eval_routing: single precision within bound (%s)', label);
    results{end, 2} = err_single < bound_single;
end

% --- Global default picked up; per-call overrides ---
rng(42, 'twister');
p = sort(1000 * rand(6, 1));
w = 0.5 + rand(6, 1);
sigma = 12;
dens = buildExpTens(p, w, sigma, 3, true, false, 0, 'verbose', false);
dens = internal.ensureExpTensExpensive(dens);
X = 1000 * rand(2, 20);
ref = local_ref_eval(dens, X);

% Default Inf: matches reference (non-periodic, so still bit-identical
% in practice, but use the same tolerance contract for consistency)
v1 = evalExpTens(dens, X, 'method', 'centres', 'verbose', false);
results{end+1, 1} = 'eval_routing: default Inf matches reference (global)';
results{end, 2} = local_within_rtol(v1, ref, 1e-12);

% Set global default = 6
mptDefaults('truncationSigmas', 6);
v2 = evalExpTens(dens, X, 'method', 'centres', 'verbose', false);
weight_mass = sum(abs(dens.wJ));
bound = max(1e-12, 10 * weight_mass * exp(-18));
results{end+1, 1} = 'eval_routing: global default 6 within bound';
results{end, 2} = max(abs(v2 - ref)) < bound;

% Per-call Inf overrides global=4
mptDefaults('truncationSigmas', 4);
v3 = evalExpTens(dens, X, 'method', 'centres', ...
    'truncationSigmas', Inf, 'verbose', false);
results{end+1, 1} = 'eval_routing: per-call overrides global';
results{end, 2} = local_within_rtol(v3, ref, 1e-12);

mptDefaults('reset');

if standalone_er
    nTests_er = size(results, 1);
    nPass_er = sum([results{:, 2}]);
    fprintf('  test_eval_routing: %d/%d passed\n', nPass_er, nTests_er);
    if nPass_er < nTests_er
        for ii = 1:nTests_er
            if ~results{ii, 2}
                fprintf('    FAIL: %s\n', results{ii, 1});
            end
        end
    end
end


function tf = local_within_rtol(actual, ref, rtol)
%LOCAL_WITHIN_RTOL  True iff max relative error <= rtol.
%   Mirrors numpy.testing.assert_allclose(actual, ref, rtol=rtol, atol=0).
%   Returns true for empty/all-zero references (vacuously).
    err = max(abs(actual(:) - ref(:)));
    denom = max(abs(ref(:)));
    if denom == 0
        tf = (err == 0);
    else
        tf = (err / denom) <= rtol;
    end
end


function v = local_ref_eval(dens, X)
%LOCAL_REF_EVAL  Frozen v2.0 evalFull body for parity reference.
%   Assumes `dens` already has its heavy fields materialised (caller
%   has invoked internal.ensureExpTensExpensive). The frozen body reads
%   the flat single-multiset layout, so present the density through the
%   view (single-multiset densities are the A = N = 1 corner of a
%   MaetDensity).
    dens = internal.singleMultisetView(dens);
    Centres = dens.Centres;
    wJ      = dens.wJ;
    sigma   = dens.sigma;
    r       = dens.r;
    dim     = dens.dim;
    isRel   = dens.isRel;
    isPer   = dens.isPer;
    J       = dens.period;
    nJ      = dens.nJ;
    nQ      = size(X, 2);

    D = reshape(Centres, dim, nJ, 1) - reshape(X, dim, 1, nQ);
    if isPer
        D = mod(D + J/2, J) - J/2;
    end
    if isRel
        Qvec = sum(D.^2, 1) - sum(D, 1).^2 / r;
    else
        Qvec = sum(D.^2, 1);
    end
    E = reshape(exp(-Qvec(:) / (2 * sigma^2)), nJ, nQ);
    v = wJ(:)' * E;
end

% Restore caller's pre-test defaults eagerly when run
% standalone (fires the helper's onCleanup destructor on
% script exit; guarded so we don't clear a like-named
% variable when this file was run from test_mpt.m, where
% the standalone branch was skipped).
if exist('cleanupDefaults_er', 'var')
    clear cleanupDefaults_er
end
